File: wibb-android/app/src/main/java/com/example/wibb/data/Offer.java
package com.example.wibb.data;

import java.time.LocalDate;

public class Offer {
    private Store store;
    private Brand brand;
    private int price;
    private LocalDate start;
    private LocalDate end;

    public Offer() {

    }

    public Store getStore() {
        return store;
    }

    public void setStore(Store store) {
        this.store = store;
    }

    public Brand getBrand() {
        return brand;
    }

    public void setBrand(Brand brand) {
        this.brand = brand;
    }

    public int getPrice() {
        return price;
    }

    public void setPrice(int price) {
        this.price = price;
    }

    public LocalDate getStart() {
        return start;
    }

    public void setStart(LocalDate start) {
        this.start = start;
    }

    public LocalDate getEnd() {
        return end;
    }

    public void setEnd(LocalDate end) {
        this.end = end;
    }

    public boolean isValid(){
        return price >= 0 && store != null && brand != null && end != null;
    }
}

File: wibb-android/app/src/main/java/com/example/wibb/data/Offer.kt
package com.example.wibb.data

import org.json.JSONObject
import java.time.*
import java.time.format.DateTimeFormatter

class Offer {
    var store: Store? = null
    var beer: Beer? = null
    var price: Int = -1
    var startDate: LocalDate? = null
    var endDate: LocalDate? = null

    val isValid: Boolean
        get() = price >= 0 && store != null && beer != null && endDate != null

    companion object {
        fun fromJSON(joffer: JSONObject): Offer {
            val o = Offer()
            o.beer = Beer.fromJSON(joffer.getJSONObject("beer"))
            o.store = Store.fromJSON(joffer.getJSONObject("store"))
            val zonedStart = ZonedDateTime.parse(joffer.getString("startDate"), DateTimeFormatter.ISO_DATE_TIME)
            val zonedEnd = ZonedDateTime.parse(joffer.getString("endDate"), DateTimeFormatter.ISO_DATE_TIME)
            o.startDate = if (joffer.isNull("startDate")) null else zonedStart.withZoneSameInstant(ZoneId.systemDefault()).toLocalDate()
            o.endDate = if (joffer.isNull("endDate")) null else zonedEnd.withZoneSameInstant(ZoneId.systemDefault()).toLocalDate()
            o.price = joffer.getInt("price")
            return o
        }
    }

    fun toJSON(): JSONObject{
        val jo = JSONObject()
        jo.put("beer", this.beer?.toJSON())
        jo.put("store", this.store?.toJSON())
        jo.put("startDate", this.startDate?.atStartOfDay(ZoneId.systemDefault())?.toOffsetDateTime()?.format(DateTimeFormatter.ISO_DATE_TIME))
        jo.put("endDate", this.endDate?.atTime(LocalTime.MAX)?.atZone(ZoneId.systemDefault())?.toOffsetDateTime()?.format(DateTimeFormatter.ISO_DATE_TIME))
        jo.put("price", this.price)
        return jo
    }
}
